src/lintsampler/lintsampler.py

```diff
@@ -0,0 +1,535 @@
+        # or in 1d case, a list of un-linked grids
+        # i.e. cells = [np.linspace(-12,0,100),np.linspace(2,12,100)]
+        # or a list of grids that have been preconstructed and stacked,
+        # i.e. [np.stack(np.meshgrid(np.linspace(-12,0,100),np.linspace(-4,0,50), indexing='ij'), axis=-1),np.stack(np.meshgrid(np.linspace(0,12,100),np.linspace(0,4,50), indexing='ij'), axis=-1)]
+        elif isinstance(cells,list):
+            self.eval_type = 'freesample'
+
+            # variable controlling whether grids have already been constructed or not
+            _gridsconstructed = False
+
+            # how many grid boundaries are we being handed?
+            self.ngrids = len(cells)
+
+            # set a dummy dimension for case checking (also relevant for debugging?)
+            self.dim = 0
+
+            # check if we are being passed grids already
+            if isinstance(cells[0],np.ndarray):
+                _gridsconstructed = True
+
+                # check special case dimensionality: we are being passed multiple 1d grids
+                if (cells[0].shape[0] == cells[0].shape[-1]):
+                    self.dim = 1
+
+            # infer the dimensionality: this will work whether the input is a single cell or list of cells
+            # only proceed with determination if not 1d case
+            if (self.dim == 0):
+                if _gridsconstructed:
+                    self.dim = cells[0].shape[-1]
+                else:
+                    self.dim = len(cells[0])
+
+            # create holding arrays
+            grids                 = [] # the grids constructed from the inputs (no need to expose)
+            self.gridshape        = [] # the shape of the input grids
+            self.ngridentries     = [] # the number of entries in the grid
+            self.nedgegridentries = [] # the number of entries in the edge grid
+            self.edgearrays       = [] # the flattened grids, each [ngridentries,dim]
+
+            # loop through the boundaries, construct the grids, and flatten
+            for ngrid in range(0,self.ngrids):
+               
+                if _gridsconstructed:
+                    # check that the grid has the same dimensionality as the others
+                    if ((self.dim>1) & (cells[ngrid].shape[-1]!=self.dim)):
+                        raise ValueError("LintSampler._setgrid: All input grids must have the same dimensionality.")
+                    
+                    grid = cells[ngrid]
+
+                else:
+                    # check that the dimensions are properly specified for the grid
+                    if len(cells[ngrid])!=self.dim:
+                        raise ValueError("LintSampler._setgrid: All input cells must have the same dimensionality.")
+
+                    # construct the grid
+                    grid = np.stack(np.meshgrid(*cells[ngrid], indexing='ij'), axis=-1)
```

Comment:
This fixes a couple of issues:
- skips the `meshgrid` business in the 1D case
- previously some types of `cells` input were going down the wrong track

```suggestion
            # infer dimensionality
            # cells = list of tuples of 1D iterables
            if isinstance(cells[0], tuple) and hasattr(cells[0][0], "__len__"):
                self.dim = len(cells[0])
            # cells = list of (k+1)-d grids (i.e. preconstructed k-d grids)
            elif isinstance(cells[0], np.ndarray) and cells[0].ndim > 1:
                _gridsconstructed = True
                self.dim = cells[0].shape[-1]
            # cells = list of 1d iterables
            elif hasattr(cells[0], "__len__") and not hasattr(cells[0][0], "__len__"):
                self.dim = 1
            else:
                raise ValueError("Input cells do not make sense.")

            # create holding arrays
            grids                 = [] # the grids constructed from the inputs (no need to expose)
            self.gridshape        = [] # the shape of the input grids
            self.ngridentries     = [] # the number of entries in the grid
            self.nedgegridentries = [] # the number of entries in the edge grid
            self.edgearrays       = [] # the flattened grids, each [ngridentries,dim]

            # loop through the boundaries, construct the grids, and flatten
            for ngrid in range(0,self.ngrids):
               
                if _gridsconstructed:
                    # check that the grid has the same dimensionality as the others
                    if ((self.dim>1) & (cells[ngrid].shape[-1]!=self.dim)):
                        raise ValueError("LintSampler._setgrid: All input grids must have the same dimensionality.")
                    
                    grid = cells[ngrid]

                else:
                    # check that the dimensions are properly specified for the grid
                    if self.dim > 1 and len(cells[ngrid])!=self.dim:
                        raise ValueError(f"LintSampler._setgrid: All input cells must have the same dimensionality.")

                    # construct the grid
                    if self.dim > 1:
                        grid = np.stack(np.meshgrid(*cells[ngrid], indexing='ij'), axis=-1)
                    else:
                        grid = np.array(cells[ngrid])
```

Reply:
I usually like to give the package and function name when returning an error, simply because it tends to make debugging easier. But I like this handling of the inputs, much more smooth with the attribute checking!